### DISCUSSION TITLE
Custom colors stubbornly refuse to compile

### DISCUSSION THREAD

**Discussion by webhype** [tailwindlabs/tailwindcss#10428] [Help] [answered]
Title says it all -- for example `bg-primary-900` is unavailable on my React (v17) / NextJS (v11) pages. Note that `theme.extend.container` works, but the custom colors are not compiled. I don't think this is a TW bug but I don't know where else to look any more for possible failure reasons.

`package.json`:
```
    "tailwindcss": "^3.2.4",
    "@tailwindcss/typography": "^0.5.9",
    "postcss": "8.4.21",
    "postcss-flexbugs-fixes": "^5.0.2",
    "postcss-nested": "^6.0.0",
    "postcss-preset-env": "^8.0.0",
```
```
// tailwind.config.js
module.exports = {
	content: [
		"./components/**/*.{html,js,ts,jsx,tsx}", "./pages/**/*.{html,js,ts,jsx,tsx}",
	],
	safelist: [
		"dark",
		"bg-primary-900",
		{
			pattern: /(text|bg|border)-(primary|secondary|gray|red|green|blue)-(50|100|200|300|400|500|600|700|800|900)/,
			variants: [
				//	'sm', 'md', 'lg', 'xl', 'hover', 'focus', 'dark', "dark:hover", "dark:focus",
				//'lg:hover'	// add combinations and permutations
			],
		},
	], //specific classes
	blocklist: [],
	darkMode: "class", // toggle manually
	theme: {
		extend: {
			colors: {
				// See https://tailwindcss.com/docs/customizing-colors
				primary: "#ff0000",	// colors.green,
				secondary: "#0000ff", 	// colors.indigo,
			},
			container: {
				center: true,
				padding: {
					DEFAULT: "1rem",
					md: "2rem",
				},
			},
		},
		screens: {
			sm: "640px",
			md: "768px",
			lg: "1024px",
			xl: "1280px",
		},
	},
}
```

```
// postcss.config.js
module.exports = {
	plugins: {
		"postcss-import": {},
		"postcss-flexbugs-fixes": {},
		//    "postcss-nesting": {},
		"postcss-custom-properties": {},
		"tailwindcss/nesting": "postcss-nesting",
		"tailwindcss/nesting": {},
		"tailwindcss": {},
		"autoprefixer": {},
			"postcss-preset-env": {
				features: { "nesting-rules": false },
			  },
	},
}

```
What I have tried myself:

- "Use" the `bg-primary-900` class in a dummy.html file to make sure it's included
- Verified that I am not "algorithmically constructing" `bg-primary-900` or any other class name
- Various orders of things in `postcss.config.js`
- Deleted `.next`, any and all temporary build or cache folders, `node_modules`, `yarn.lock` and reinstalled everything

At my wit's end here... where else can I look?

Note 1: Also getting the dreaded error "The `purge`/`content` options have changed in Tailwind CSS v3.0." when I rebuild the project but not on minor updates (changes to html/js files).

Note 2: This all worked just fine under Tailwind 2.6, only started breaking after upgrading to 3.x

**Comment by adamwathan** [comment]
Hey! Can you create a project that reproduces the issue? Could figure this out in a minute or two for sure if I can just clone a project and play with it.

**Reply by webhype** [reply]
Thanks for the offer; unfort the whole project is rather big and convoluted and in the end it was that I imported the colors from the wrong file; see answer below.

**Comment by wongjn** [comment, marked as answer]
`primary-900` or `primary.900` does not seem to be a valid color. To make it a valid color, you'd need to do something like:
```diff
 colors: {
   // See https://tailwindcss.com/docs/customizing-colors
-  primary: "#ff0000",	// colors.green,
+  primary: {
+    900: "#ff0000",
+  },
   secondary: "#0000ff", 	// colors.indigo,
 },
 ```

**Reply by webhype** [reply]
Thanks @wongjn you helped me solve it! It's clear now that the variants 900, 800, 700 don't magically generate themselves, but you need to define them as you showed.

My end game was to use default colors like `colors.indigo`, which also no longer worked after upgrading from v2 to v3.

So I figured out that the default colors were apparently moved into their own file.

Old / v2 way that worked for me before upgrade to v3 for some reason:
```
const {colors} = require("tailwindcss/defaultTheme");
// colors.indigo contains all expected variations
```
Proper v2 and v3 way, according to the documentation:
```
const colors = require("tailwindcss/colors"); // import the colors file!
// colors.indigo contains all expected variations
```
I am not entirely clear why `colors.indigo` was available to me using the above "old way", as that's not the documented way.

Anyway, I hope this helps someone, the colors live in `tailwindcss/defaultTheme` and no-where else.
